### PR TITLE
IGNITE-17685 Sql. SQL engine should send messages asynchronously

### PR DESCRIPTION
**Pull request by korlov42**
The patch is split onto two commits. The first one introduces dead code elimination and minor refactoring like moving sending of close message from ExchangeService. The second commit introduce actual refactoring of MessageService

### CODE REVIEW

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 1]
```diff
@@ -467,15 +467,22 @@ private void sendFragment(
...
+            );
 
-                throw ex;
+            try {
+                return messageService.send(targetNodeName, request);
+            } catch (Throwable th) {
+                // it's not expected MessageService to throw any exception, yet it may be possible
+                // due to unpredictable side effects of different implementations. To avoid blocking
+                // of node on stopping, let's complete initialization future.
+                remoteFragmentInitializationCompletionFuture.complete(null);
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:483]: Should we also remove future from remoteFragmentInitCompletion map if exception is re-thrown.

**Review by AMashenkov — APPROVED** [review]

**Review by lowka — APPROVED** [review]

**Review by zstan — APPROVED** [review]
looks good, besides minor comments

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 2]
```diff
@@ -543,21 +550,19 @@ private void executeFragment(IgniteRel treeRoot, ExecutionContext<RowT> ectx) {
...
-            try {
-                messageService.send(
-                        origNodeName,
-                        FACTORY.queryStartResponse()
-                                .queryId(ectx.queryId())
-                                .fragmentId(ectx.fragmentId())
-                                .build()
-                );
-            } catch (IgniteInternalCheckedException e) {
-                throw new IgniteInternalException(MESSAGE_SEND_ERR, "Failed to send reply. [nodeName=" + origNodeName + ']', e);
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java:567]: removed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java [hunk 3]
```diff
@@ -467,15 +467,22 @@ private void sendFragment(
...
+            remoteFragmentInitCompletion.put(
+                    new RemoteFragmentKey(targetNodeName, fragment.fragmentId()),
+                    remoteFragmentInitializationCompletionFuture
+            );
 
-                throw ex;
+            try {
+                return messageService.send(targetNodeName, request);
+            } catch (Throwable th) {
+                // it's not expected MessageService to throw any exception, yet it may be possible
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java]: > additionally for now it throws NODE_LEFT_ERR

this is not how it expected to be, so thanks to pointing out!

> i think we don`t need to write about expectations

Although it's implementation defined, every implementation must conform the contract, which defined by MessageService interface. Currently, method `send` declares no exception in `throws` section. Besides, it returns future representing result of operation. Thus, I think it's ok in that particular case to expected of this method not to throw any exception.

Anyway, I reworked the test that make me put this try-catch block (and also fix missed NODE_LEFT_ERR), so we do not need it anymore.

**Pull request merged**